### PR TITLE
Connect progress tracking dataclasses to loops

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

See title.

Part of #8131
Includes #8265 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [n/a] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 1]
```diff
@@ -47,13 +49,31 @@ def __init__(self) -> None:
...
+        optim_progress: Optional[OptimizationProgress] = None,
+        **kwargs: Any
+    ) -> None:
+        """Connects the loop with necessary arguments like the trainer"""
+        super().connect(trainer, *args, **kwargs)
+        if progress is not None:
+            self.progress = progress
+        if optim_progress:
+            self.optim_progress = optim_progress
+
```
Comment (ananthsub) [inline pytorch_lightning/loops/batch/training_batch_loop.py:76]: could you describe why progress/optim_progress are passed through connect?

File: pytorch_lightning/loops/dataloader/evaluation_loop.py [hunk 2]
```diff
@@ -31,11 +32,15 @@ class EvaluationLoop(DataLoaderLoop):
 
     def __init__(self):
         super().__init__()
-        self._max_batches: Optional[Union[int, Sequence[int]]] = None
         self.outputs = []
+
         self.epoch_loop = EvaluationEpochLoop()
+
+        self.results = ResultCollection(training=False)
```
Comment (ananthsub) [inline pytorch_lightning/loops/dataloader/evaluation_loop.py:39]: are results meant to be mutable for external callers beyond the logger connector?

Comment (carmocca) [reply pytorch_lightning/loops/dataloader/evaluation_loop.py]: Nope, removed the change

File: pytorch_lightning/loops/batch/training_batch_loop.py (hunk 1 above)
Comment (carmocca) [reply pytorch_lightning/loops/batch/training_batch_loop.py:75]: We need to connect the progress instances between loops, just as we do for the trainer instance.

This is what is checked in the test added

File: pytorch_lightning/loops/dataloader/evaluation_loop.py [hunk 3]
```diff
@@ -31,11 +32,14 @@ class EvaluationLoop(DataLoaderLoop):
...
-        self._max_batches: Optional[Union[int, Sequence[int]]] = None
         self.outputs = []
+        self.progress = EpochLoopProgress()
+
         self.epoch_loop = EvaluationEpochLoop()
-        self._has_run: bool = False
+
         self._results = ResultCollection(training=False)
+        self._max_batches: Optional[Union[int, Sequence[int]]] = None
+        self._has_run: bool = False
```
Comment (tchaton) [inline pytorch_lightning/loops/dataloader/evaluation_loop.py:42]: Note: Let's remove this hack for progress tracking instead.

File: pytorch_lightning/trainer/trainer.py [hunk 4]
```diff
@@ -347,10 +348,10 @@ def __init__(
         self.validate_loop = EvaluationLoop()
         self.test_loop = EvaluationLoop()
         self.predict_loop = PredictionLoop()
-        self.fit_loop.connect(self)
-        self.validate_loop.connect(self)
-        self.test_loop.connect(self)
-        self.predict_loop.connect(self)
+        self.fit_loop.connect(self, progress=FitLoopProgress())
```
Comment (tchaton) [inline pytorch_lightning/trainer/trainer.py:351]: Perfect !

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 5]
```diff
@@ -67,11 +69,19 @@ def done(self) -> bool:
...
+        progress: Optional[TrainingEpochProgress] = None,
+        **kwargs: Any
+    ) -> None:
+        """Connects the loop with necessary arguments like the trainer"""
         super().connect(trainer, *args, **kwargs)
-        self.batch_loop.connect(trainer)
-        self.val_loop.connect(trainer)
+        if progress is not None:
+            self.progress = progress
+        self.batch_loop.connect(trainer, progress=self.progress.batch, optim_progress=self.progress.optim)
```
Comment (carmocca) [inline pytorch_lightning/loops/epoch/training_epoch_loop.py:83]: There's a rough edge here for the optimization progress

On the one hand, we track progress at the epoch level, since we don't care much about the number of optimizer steps etc performed per batch

On the other hand, it is the batch loop who needs to update the counters.

So I would prefer to not pass it down to the next loop here, however, the batch loop would then need to somehow notify the parent loop to update this progress or both should have references to it.

Thoughts?

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 6]
```diff
@@ -48,13 +50,30 @@ def __init__(self) -> None:
         self.running_loss: TensorRunningAccum = TensorRunningAccum(window_length=20)
         self.batch_idx: int = 0
         self.split_idx: Optional[int] = None
-        self._warning_cache: WarningCache = WarningCache()
+        self.progress = BatchProgress()
+        self.optim_progress = OptimizationProgress()
```
Comment (awaelchli) [inline pytorch_lightning/loops/batch/training_batch_loop.py:54]: great!
Let's say we would break out the optimizer loop into it's own loop class. 
We would then simply move this property to that optimizer loop?

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 7]
```diff
@@ -48,13 +50,30 @@ def __init__(self) -> None:
...
+        *args: Any,
+        progress: Optional[BatchProgress] = None,
+        optim_progress: Optional[OptimizationProgress] = None,
+        **kwargs: Any
+    ) -> None:
+        """Connects the loop with necessary arguments like the trainer"""
+        super().connect(trainer, *args, **kwargs)
+        if progress is not None:
+            self.progress = progress
+        if optim_progress:
```
Comment (awaelchli) [inline pytorch_lightning/loops/batch/training_batch_loop.py:74]:
```suggestion
        if progress is not None:
            self.progress = progress
        if optim_progress is not None:
```

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/loops/batch/training_batch_loop.py (hunk 6 above)
Comment (carmocca) [reply pytorch_lightning/loops/batch/training_batch_loop.py:54]: Depends on the responsabilities of the optimizer loop but likely yes.

Do you have something more specific in mind?

Comment (awaelchli) [reply pytorch_lightning/loops/batch/training_batch_loop.py:54]: just wanted to confirm we have the flexibility available. 
what we would track as progress in such an optimizer loop is not yet so clear but we will see :)

File: pytorch_lightning/loops/epoch/training_epoch_loop.py (hunk 5 above)
Comment (awaelchli) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:83]: I would avoid adding references to the "parent" loop as long as possible. Such connections could potentially add a lot of confusion when it comes to control flow and nesting, and for debugging. 


> On the other hand, it is the batch loop who needs to update the counters.

which loop exactly is the one that would update the .batch.processed counter? I would say the TrainingEpochLoop, since this is the one who advances from one batch to the next. Don't you think?

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/loops/epoch/training_epoch_loop.py (hunk 5 above)
Comment (carmocca) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:83]: > which loop exactly is the one that would update the .batch.processed counter? I would say the TrainingEpochLoop, since this is the one who advances from one batch to the next. Don't you think?

Given the initial definition for batch trackers and the current loop structure:

ready: batch fetched from dataloader
started: after on_train_batch_start
processed: after training_step
completed: after on_train_batch_end

Means `ready/completed` would be updated in `TrainingEpochLoop` and `started/processed` in `TrainingBatchLoop`

Comment (awaelchli) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:83]: what prevents us from updating the batch progress consistently all in one loop?

Comment (awaelchli) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:83]: we can discuss further on the integration PR
